### PR TITLE
sjawhar-legion-415: refactor(daemon): replace env vars with CLI flags

### LINKED ISSUE #415: refactor(daemon): replace env vars with CLI flags — discoverable, self-documenting config
## Problem

The daemon uses env vars for configuration that are impossible to discover without reading the code. Sami's directive: CLI flags as the primary interface, env vars as fallbacks only.

## Scope — Audit and Convert ALL Env Vars

| Current Env Var | New CLI Flag | Notes |
|----------------|-------------|-------|
| `LEGION_ID` | `--projects` | Unified list, no primary/extra split |
| `LEGION_EXTRA_PROJECTS` | folded into `--projects` | Removed entirely |
| `LEGION_ISSUE_BACKEND` | `--backend` | Already exists as `-b` |
| `LEGION_RUNTIME` | `--runtime` | Already exists as `-r` |
| `LEGION_DAEMON_PORT` | `--port` | |
| `LEGION_CONTROLLER_SESSION_ID` | `--controller-session` | |
| `LEGION_SHORT_ID` | derive from projects | May not need a flag |
| `ENVOY_URL` | `--envoy-url` | |
| `OPENCODE_AUTO_HEAP_SNAPSHOT` | `--no-heap-snapshot` | Boolean flag |
| `OPENCODE_FORCE_GC` | `--force-gc` | Boolean flag |
| Any others found in codebase | TBD | Audit needed |

## Design

- CLI flags are the primary interface: `legion start --projects trajectory-labs-pbc/2,sjawhar/2 --backend github --runtime opencode`
- `legion start --help` shows all options (self-documenting)
- Env vars still work as fallbacks (for CI/containers) but are not the documented interface
- **Convention going forward: no new env vars** — all new config via CLI flags

## Context

- #92 introduced `LEGION_EXTRA_PROJECTS` env var — this subsumes that
- Sami flagged the primary/extra split as messy, then escalated to "stop using env vars entirely"
- citty (the CLI framework) already supports this pattern well

### PR DESCRIPTION
**Pull request by sjawhar**
Closes #415

## Summary

Replace `legion start` daemon env-var configuration with discoverable CLI flags while preserving env fallbacks and the controller subprocess env contract.

### Changes
- **CLI-aware config resolution**: `loadConfig(env, cliArgs)` with strict `CLI > env > default` precedence
- **New `legion start` flags**: `--projects`, `--port`, `--controller-session`, `--envoy-url` (in addition to existing `-b`, `-r`, `-w`, `-p`)
- **No more env mutation**: `cmdStart()` no longer sets `process.env.LEGION_ISSUE_BACKEND` or `process.env.LEGION_RUNTIME`
- **Centralized Envoy URL**: All 6 `process.env.ENVOY_URL` reads in daemon/server replaced with config-driven `envoyUrl` parameter threading
- **Controller env propagation**: `buildControllerEnv()` now includes `ENVOY_URL` so spawned sessions see the CLI-overridden value

### Files (11 changed, +490 -76)
- `config.ts`: `CliArgs` interface, `envoyUrl` on `DaemonConfig`, `parseProjects()`, `coalesceValue()`
- `cli/index.ts`: `StartOptions` extended, `buildStartCliArgs()`, env mutation removed
- `daemon/index.ts`: `DaemonOverrides.cliArgs`, Envoy functions accept `envoyUrl` param
- `daemon/server.ts`: `ServerOptions.envoyUrl`, all Envoy helpers parameterized
- Tests: 490+ lines of new coverage across config, CLI, daemon index, server, integration

### Verification
- 1170 daemon tests pass (0 fail)
- Biome lint clean
- tsc clean
- `grep process.env.ENVOY_URL` returns zero hits in daemon code

### CODE REVIEW

**Comment by sjawhar** [comment]
## Behavioral Test Results

### Summary
**PASS** — 9/9 acceptance criteria verified. All daemon-scoped tests pass (827/828 — 1 pre-existing flaky test in `serve-manager.test.ts` unrelated to this PR). Biome lint clean. No `process.env.ENVOY_URL` reads remain in daemon production code.

### Results
| Criterion | Status | Evidence |
|-----------|--------|----------|
| CLI flags as primary interface (`--projects`, `--port`, `--controller-session`, `--envoy-url`) | ✅ | `startCommand.args` in `cli/index.ts:902-943` defines all flags. Verified by test: "start command exposes config flags for daemon startup" in `cli/__tests__/index.test.ts:331-350` |
| `--projects` unified list (first=primary, rest=extra, dedup) | ✅ | `parseProjects()` in `config.ts:101-135`. Tested: "parses --projects with first as legionId", "deduplicates", "CLI --projects overrides both LEGION_ID and LEGION_EXTRA_PROJECTS" in `config.test.ts:326-351` |
| `--backend`/`-b` and `--runtime`/`-r` replace env vars | ✅ | `coalesceValue(cliArgs.backend, env.LEGION_ISSUE_BACKEND)` in `config.ts:170-176`. Tested: "prefers CLI args over env vars" in `config.test.ts:262-291` |
| `--port` replaces `LEGION_DAEMON_PORT` | ✅ | `coalesceValue(cliArgs.port, env.LEGION_DAEMON_PORT)` in `config.ts:195-198`. Note: `allocatePort()` in `startDaemon()` overwrites this — **pre-existing behavior** identical to the env var (documented in implement handoff `openQuestions`) |
| `--controller-session` replaces env var | ✅ | `coalesceValue(cliArgs.controllerSession, env.LEGION_CONTROLLER_SESSION_ID)` in `config.ts:158-159`. Tested: "uses cliArgs overrides when building controller env" in `index.test.ts:1288-1345` |
| `--envoy-url` replaces `ENVOY_URL` with full centralization | ✅ | `coalesceValue(cliArgs.envoyUrl, env.ENVOY_URL, DEFAULT_ENVOY_URL)` in `config.ts:199-200`. All 6 daemon `process.env.ENVOY_URL` reads replaced. Verified by `grep -rn process.env.ENVOY_URL daemon/index.ts daemon/server.ts` → 0 matches |
| Env vars still work as fallbacks | ✅ | `loadConfig(env, cliArgs={})` path tested: "falls back to env vars when CLI args are omitted" in `config.test.ts:293-314` |
| No `process.env` mutation in `cmdStart()` | ✅ | Old `process.env.LEGION_ISSUE_BACKEND = ...` and `process.env.LEGION_RUNTIME = ...` removed (visible in diff). Tested: "buildStartCliArgs maps start options to CliArgs without mutating process.env" in `cli/__tests__/index.test.ts:1461-1503` |
| `buildControllerEnv()` emits `ENVOY_URL` | ✅ | `config.envoyUrl` propagated to controller subprocess via `buildControllerEnv()` at `index.ts:196`. Tested: "uses config envoyUrl for controller envoy subscriptions" in `index.test.ts:331-391` |

### Test Quality Critique
- **Meaningful coverage**: 92 new test lines in `config.test.ts`, 59 in `cli/__tests__/index.test.ts`, 120 in `daemon/__tests__/index.test.ts`, 40 in `server.test.ts` — testing real precedence logic, not just existence.
- **No excessive mocking**: The `index.test.ts` Envoy test spins up a real `Bun.serve()` mock server and verifies actual fetch URLs hit the configured endpoint, not `process.env`. The server test does the same for worker subscribe/unsubscribe paths.
- **Edge cases covered**: Deduplication, malformed `--projects` entries, empty CLI args fallback, `envoyUrl` default when neither CLI nor env provides one.
- **P2: Missing `--port` effectiveness test**: No test verifies that `--port` actually controls the bound port. This is because `allocatePort()` overwrites it — a pre-existing design issue, not introduced by this PR.

### Documentation Feedback
- No README or documentation files were changed in this PR. The CLI flags are self-documenting via `legion start --help`, which is the stated design goal. The issue body and plan comment serve as the design record.
- The AGENTS.md still references env vars for controller config — updating it to mention CLI flags would be a nice follow-up but is not required for this refactor.

### Observations
- **Pre-existing flaky test**: `serve-manager.test.ts > waitForHealthy > retries until healthy` — expects 3 retry calls but sometimes gets 4 due to timing sensitivity. Not in this PR's changed files.
- **Pre-existing tsc warning**: `tsconfig.json` uses deprecated `baseUrl` option (TS 6.0 deprecation). Not related to this PR.
- **Pre-existing test failures**: 2 tests fail in `packages/envoy/infra/` due to missing `@pulumi/docker` module — infrastructure dependency, not related.
- **`--port` semantic gap**: The flag is parsed correctly at the config layer but `startDaemon()` → `allocatePort()` overwrites it. The implement handoff correctly documents this as pre-existing behavior identical to `LEGION_DAEMON_PORT` env var. Worth a follow-up issue.
- **ENVOY_URL in envoy-plugin**: `packages/envoy-plugin/src/index.ts` still reads `process.env.ENVOY_URL` directly, but this is outside the daemon package scope. PR body correctly scopes the claim to "daemon/server" code.

### Config Compliance
- require_specific_task: false — N/A
- require_taiga_evidence: false — N/A

**Comment by sjawhar** [comment]
## Review Summary

**CRITICAL (P1):** 0 issues
**IMPORTANT (P2):** 0 issues
**MINOR (P3):** 3 suggestions

**Verdict: Approved to merge.** Clean refactor that correctly threads CLI args through the daemon startup chain while preserving env fallbacks and the controller subprocess env contract.

---

### Acceptance Criteria Verification

All 8 env vars from issue #415 are covered:
- `LEGION_ID` → `--projects` (first entry = primary) ✅
- `LEGION_EXTRA_PROJECTS` → folded into `--projects` ✅
- `LEGION_ISSUE_BACKEND` → `--backend` / `-b` ✅
- `LEGION_RUNTIME` → `--runtime` / `-r` ✅
- `LEGION_DAEMON_PORT` → `--port` ✅
- `LEGION_CONTROLLER_SESSION_ID` → `--controller-session` ✅
- `LEGION_SHORT_ID` → derived from `legionId.slice(0, 8)` ✅
- `ENVOY_URL` → `--envoy-url` ✅
- `OPENCODE_AUTO_HEAP_SNAPSHOT`, `OPENCODE_FORCE_GC` → correctly dropped (not in codebase)

### Key Design Decisions (verified correct)

1. **`buildControllerEnv()` now emits `ENVOY_URL`** — Not in original plan but required for `--envoy-url` to propagate to controller subprocess. Correct deviation.
2. **Empty-string backend/runtime handling** — The `backendValue`/`runtimeValue` fallback preserves the original error-on-empty-string behavior through `coalesceValue` which skips empty strings. Correct but complex (see P3 below).
3. **Fire-and-forget Envoy pattern preserved** — All parameterized Envoy functions maintain the non-blocking pattern per `docs/solutions/daemon/envoy-auto-subscription-patterns.md`.

### P3 Findings

1. **Empty-string backend/runtime fallback chain** (config.ts:171-172, 179-180) — The `backendValue = rawBackend ?? (cliArgs.backend === "" ? "" : ...)` logic is correct but hard to read. A brief inline comment explaining why `coalesceValue` alone doesn't suffice (it skips empty strings, but validation needs to distinguish explicit empty from absent) would help future maintainers.

2. **`--port` flag accepted but overwritten** — `allocatePort()` in `startDaemon()` overwrites `config.daemonPort`. This is pre-existing behavior with `LEGION_DAEMON_PORT`, not introduced by this PR. The implementer correctly documented this as an open question.

3. **Remaining non-CLI env vars** — `OPENCODE_MAX_RSS_GB`, `OPENCODE_RSS_CHECK_INTERVAL`, `LEGION_FEEDBACK_DISABLED`, `LEGION_FEEDBACK_MAX_BYTES` are still env-only. The plan explicitly scoped these out. Consider a follow-up if the "no new env vars" convention should extend to converting these.

### Evidence
- CI status: No checks reported on branch
- `grep process.env.ENVOY_URL packages/daemon/src/daemon/index.ts packages/daemon/src/daemon/server.ts` returns zero hits
- Test coverage: 490+ lines of new tests across config, CLI, daemon index, server
- Implementer verification: 1170 daemon tests pass, Biome clean, tsc clean

File: packages/daemon/src/daemon/config.ts [hunk 1]
```diff
@@ -122,19 +167,23 @@ export function loadConfig(env: NodeJS.ProcessEnv = process.env): DaemonConfig {
 
   validateControllerPrompt(controllerPrompt);
 
-  const rawBackend = env.LEGION_ISSUE_BACKEND;
-  if (rawBackend !== undefined && rawBackend !== "linear" && rawBackend !== "github") {
-    throw new Error(`LEGION_ISSUE_BACKEND must be 'linear' or 'github' (got: ${rawBackend})`);
+  const rawBackend = coalesceValue(cliArgs.backend, env.LEGION_ISSUE_BACKEND);
+  const backendValue =
+    rawBackend ?? (cliArgs.backend === "" ? "" : env.LEGION_ISSUE_BACKEND === "" ? "" : undefined);
```
Comment (sjawhar) [inline packages/daemon/src/daemon/config.ts:172]: **[P3/MINOR]:** This fallback chain is correct but hard to reason about. `coalesceValue()` skips empty strings by design, but the original code threw errors on empty-string backend/runtime. The `backendValue = rawBackend ?? (cliArgs.backend === "" ? "" : ...)` reconstructs that behavior. A brief inline comment (e.g. `// coalesceValue skips empty strings; reconstruct them for validation`) would help future readers.

**Comment by sjawhar** [comment]
Superseded by #436 (PR #442) — config file replaces CLI flags as the end state.
